source/extensions/filters/http/aws_lambda/sts_connection_pool.cc

```diff
@@ -0,0 +1,217 @@
+    fetcher_->cancel();
+  }
+};
+
+void StsConnectionPoolImpl::init(const envoy::config::core::v3::HttpUri &uri,
+                                 const absl::string_view web_token) {
+  request_in_flight_ = true;
+  fetcher_->fetch(uri, role_arn_, web_token, this);
+}
+
+StsConnectionPool::Context *
+StsConnectionPoolImpl::add(StsConnectionPool::Context::Callbacks *callbacks) {
+  ContextImpl *ctx_ptr{new ContextImpl(callbacks, *this, dispatcher_)};
+  std::unique_ptr<ContextImpl> ctx{ctx_ptr};
+
+  LinkedList::moveIntoList(std::move(ctx), connection_list_);
+  return ctx_ptr;
+};
+
+void StsConnectionPoolImpl::onSuccess(const absl::string_view body) {
+  ASSERT(body != nullptr);
+  request_in_flight_ = false;
+
+// using a macro as we need to return on error
+// TODO(yuval-k): we can use string_view instead of string when we upgrade to
+// newer absl.
+#define GET_PARAM(X)                                                           \
+  std::string X;                                                               \
+  {                                                                            \
+    std::match_results<absl::string_view::const_iterator> matched;             \
+    bool result = std::regex_search(body.begin(), body.end(), matched,         \
+                                    StsResponseRegex::get().regex_##X);        \
+    if (!result || !(matched.size() != 1)) {                                   \
+      ENVOY_LOG(trace, "response body did not contain " #X);                   \
+      onFailure(CredentialsFailureStatus::InvalidSts);                         \
+      return;                                                                  \
+    }                                                                          \
+    const auto &sub_match = matched[1];                                        \
+    decltype(X) matched_sv(sub_match.first, sub_match.length());               \
+    X = std::move(matched_sv);                                                 \
+  }
+
+  GET_PARAM(access_key);
+  GET_PARAM(secret_key);
+  GET_PARAM(session_token);
+  GET_PARAM(expiration);
+
+  SystemTime expiration_time;
+  absl::Time absl_expiration_time;
+  std::string error;
+  if (absl::ParseTime(absl::RFC3339_sec, expiration, &absl_expiration_time,
+                      &error)) {
+    ENVOY_LOG(trace, "Determined expiration time from STS credentials result");
+    expiration_time = absl::ToChronoTime(absl_expiration_time);
+  } else {
+    expiration_time = api_.timeSource().systemTime() + REFRESH_STS_CREDS;
+    ENVOY_LOG(trace,
+              "Unable to determine expiration time from STS credentials "
+              "result (error: {}), using default",
+              error);
+  }
+
+  StsCredentialsConstSharedPtr result = std::make_shared<const StsCredentials>(
+      access_key, secret_key, session_token, expiration_time);
+
+  // Send result back to Credential Provider to store in cache
+  callbacks_->onSuccess(result, role_arn_);
```

Comment:
onSuccess is the only function used in the callback.. maybe change the name onResult?